File: src/main/java/net/masterthought/dlanguage/features/DCommenter.java
package net.masterthought.dlanguage.features;

import com.intellij.lang.CodeDocumentationAwareCommenter;
import com.intellij.psi.PsiComment;
import com.intellij.psi.tree.IElementType;
import net.masterthought.dlanguage.psi.DlangTypes;
import org.jetbrains.annotations.Nullable;


public class DCommenter implements CodeDocumentationAwareCommenter {
    public String getLineCommentPrefix() {
        return "//";
    }

    public String getBlockCommentPrefix() {
        return "/*";
    }

    public String getBlockCommentSuffix() {
        return "*/";
    }

    public String getCommentedBlockCommentPrefix() {
        return null;
    }

    public String getCommentedBlockCommentSuffix() {
        return null;
    }

    @Nullable
    public IElementType getLineCommentTokenType() {
        return DlangTypes.LINE_COMMENT;
    }

    @Nullable
    public IElementType getBlockCommentTokenType() {
        return DlangTypes.BLOCK_COMMENT;
    }

    public String getDocumentationCommentPrefix() {
        return "/**";
    }

    public String getDocumentationCommentLinePrefix() {
        return "*";
    }

    public String getDocumentationCommentSuffix() {
        return "*/";
    }

    public boolean isDocumentationComment(final PsiComment element) {
        return false;
//        return element.getTokenType() == DElementTypes.DOCCOMMENT_LINE ||
//                element.getTokenType() == DElementTypes.DOCCOMMENT_MULTI ||
//                element.getTokenType() == DElementTypes.DOCCOMMENT_NESTED;
    }

    @Nullable
    public IElementType getDocumentationCommentTokenType() {
        return DlangTypes.LINE_COMMENT;
    }
}

File: src/main/kotlin/io/github/intellij/dlanguage/features/DCommenter.kt
package io.github.intellij.dlanguage.features

import com.intellij.lang.CodeDocumentationAwareCommenter
import com.intellij.psi.PsiComment
import com.intellij.psi.tree.IElementType
import io.github.intellij.dlanguage.psi.DlangTypes

class DCommenter : CodeDocumentationAwareCommenter {
    override fun getLineCommentPrefix(): String {
        return "//"
    }

    override fun getBlockCommentPrefix(): String {
        return "/*"
    }

    override fun getBlockCommentSuffix(): String {
        return "*/"
    }

    override fun getCommentedBlockCommentPrefix(): String? {
        return null
    }

    override fun getCommentedBlockCommentSuffix(): String? {
        return null
    }

    override fun getLineCommentTokenType(): IElementType? {
        return DlangTypes.LINE_COMMENT
    }

    override fun getBlockCommentTokenType(): IElementType? {
        return DlangTypes.BLOCK_COMMENT
    }

    override fun getDocumentationCommentPrefix(): String {
        return "/**"
    }

    override fun getDocumentationCommentLinePrefix(): String {
        return "*"
    }

    override fun getDocumentationCommentSuffix(): String {
        return "*/"
    }

    override fun isDocumentationComment(element: PsiComment): Boolean {
        return false
//        return element.getTokenType() == DElementTypes.DOCCOMMENT_LINE ||
//                element.getTokenType() == DElementTypes.DOCCOMMENT_MULTI ||
//                element.getTokenType() == DElementTypes.DOCCOMMENT_NESTED;
    }

    override fun getDocumentationCommentTokenType(): IElementType? {
        return DlangTypes.LINE_COMMENT
    }
}
